capnp/src/private/arena.rs

```diff
@@ -154,15 +155,9 @@ where
 }
 
 pub trait BuilderArena: ReaderArena {
-    // These methods all take an immutable &self because otherwise a StructBuilder<'a>
-    // would need a `&'a mut BuilderArena` and `StructBuilder::borrow()` would
-    // have lifetime issues. (If `'a: 'b`, then a `&'a (BuilderArena + 'a)` can be
-    // converted to a `&'b (BuilderArena + 'b)`, but a `&'a mut (BuilderArena + 'a)`
-    // *cannot* be converted to a `&'b mut (BuilderArena + 'b)`. See some discussion here:
-    // https://botbot.me/mozilla/rust/2017-01-31/?msg=80228117&page=19 .)
```

Comment:
Ah, so this comment captures the reason why I thought that the `RefCell` was necessary. (I updated it for `dyn` and `reborrow` here: https://github.com/capnproto/capnproto-rust/commit/5ad7849382d86e6ed8043a9b7ba3c52fbc631304.)

I wonder what changed so that this is now possible? Or maybe I was mistaken all along?

Reply:
Wow, this failed in Rust 1.62: https://godbolt.org/z/4j354zad3
and started working in Rust 1.63: https://godbolt.org/z/3hv5x3vrK

I'm a little bit worried that this means that the borrowchecker is now unsound! But if it's not, then this is good news for us, because it will let us make these methods take `&mut self` and therefore let us remove the `RefCell`.

Reply:
I think this is because NLL has been fully stabilized in 1.63.0: https://blog.rust-lang.org/2022/08/05/nll-by-default.html

Reply:
Looking at the code, I don't see any obvious reasons it should not compile or can be unsound.

Reply:
I can probably ask t-types team on zulip for a confirmation?

Reply:
Aha. Here's a superficially similar thing that (thankfully) still does not typecheck:
```rust
struct Foo<'a> {
    buf: &'a mut [u8],
}

struct Bar<'a> {
    inner : &'a mut Foo<'a>,
}

impl <'a> Bar <'a> {
    fn reborrow<'b>(&'b mut self) -> Bar<'b>  {
        Bar { inner: self.inner }
    }
}

pub fn main() {}
```

Allowing this would be unsound, because then we could do:
```rust
pub fn main() {
   let mut v1 : Vec<u8> = vec![1,2,3];
   let mut foo1 = Foo { buf: &mut v1 };
   let mut bar = Bar { inner: &mut foo1 };
   {
        let mut v2 : Vec<u8> = vec![4,5];
        let mut foo2 = Foo { buf: &mut v2 };
        *bar.reborrow().inner = foo2;
   }
   println!("{}", bar.inner.buf[0]); // accesses v2, which is out of scope!
}
```

The problem is that we can *assign* a value to `Bar::inner` that does not live long enough.

In our current case, however, we are dealing with trait objects (i.e. `dyn`), and those do *not* allow assignment through `&mut` references (because the size is unknown). So I don't see any obvious unsoundness.

Reply:
That's to say, it's perhaps understandable that rustc pre-NLL was too conservative in this case.

Reply:
Ah, and according to the referece (https://doc.rust-lang.org/reference/subtyping.html#variance),

`dyn Trait<T> + 'a` is "covariant" in `'a`. Pre-NLL it was "invariant", if I understand correctly.

Reply:
Actually, I can get my little example to typecheck on an even older rustc (e.g. 1.30) if I avoid the `..*self` syntax:
https://godbolt.org/z/8EGx4v3zK

Reply:
posted on zulip: https://rust-lang.zulipchat.com/#narrow/stream/144729-t-types/topic/subtype.20covariance.20with.20.60.26'a.20mut.20.28dyn.20R.20.2B.20'a.29.60/near/321516958

Reply:
If I understand correctly, 1.63.0 just made it work with the `..*self` syntax, but it was working before using regular field assignment. So arguably, we could have introduced such code before 1.63.0 and get rid of `RefCell` already.

So answering your first question from zulip:

> Can I rely on this continuing to typecheck? I'd like to migrate a bunch of my code to depend on this, because I had been working around it previously.

Unless there is something inherently unsound about accepting this pattern; I fail to see why the compiler would suddenly reject it. That code has been working for years and such pattern is probably used in various places; so rejecting it would be tricky.

Given all of this, do you see any blocker that would prevent merging this?

Reply:
I want to convince myself it's not unsound first. :)

I think this PR should be considered a breaking change and that we should do a version bump soon including it and some of the other stuff that's been waiting on a version bump.